src/app/services/canvas.service.ts

```diff
@@ -1230,6 +1207,72 @@ export class CanvasService {
         studentId: string,
         overrideTitlePrefix: string
     ): Promise<boolean> {
+        type OverrideDate = Date | null;
+        type OverrideDates = { unlockAt: OverrideDate; lockAt: OverrideDate; dueAt: OverrideDate };
+
+        /**
+         * Merges two override dates into a single choice.
+         * @param min (default: true) if true, return earliest date. if false, return latest date.
+         * @param preserveDate (default: false) if true, returns the other date if one is null
+         * @returns the merged date or null
+         */
+        function mergeOverrideDate(a: OverrideDate, b: OverrideDate, min = true, preserveDate = false): OverrideDate {
+            if (a == null && preserveDate) return b;
+            if (b == null && preserveDate) return a;
+            if (a == null || b == null) return null;
+            if (min) {
+                return compareDesc(a, b) == 1 ? a : b;
+            } else {
+                return compareAsc(a, b) == 1 ? a : b;
+            }
+        }
+
+        /**
+         * @returns true if both dates are equal or both are null
+         */
+        function isOverrideDateEqual(a: OverrideDate, b: OverrideDate): boolean {
+            if (a == null && b == null) return true;
+            if (a == null || b == null) return false;
+            return isEqual(a, b);
+        }
+
+        /**
+         * @returns true if each pair of dates are equal (either same date or both null)
+         */
+        function areOverrideDatesEqual(a: OverrideDates, b: OverrideDates): boolean {
+            return (
+                isOverrideDateEqual(a.unlockAt, b.unlockAt) &&
+                isOverrideDateEqual(a.dueAt, b.dueAt) &&
+                isOverrideDateEqual(a.lockAt, b.lockAt)
+            );
+        }
+
+        /**
+         * Merges unlock, due, and lock dates.
+         * Due at and Lock at use the latest date. Unlock at uses the earliest.
+         * Null is basically both +∞ and -∞.
+         *
+         * @param preserveDate If passed a boolean, will prioritize preserving dates (if true) or returning nulls (if false) for all override dates
+         * @param skipMerging configure if merging should be skipped for the specified override dates (returns the value of the `a` argument instead)
+         * @returns object with merged date results
+         */
+        function mergeOverrideDates(
+            a: OverrideDates,
+            b: OverrideDates,
+            preserveDate?: boolean,
+            skipMerging?: { [key in keyof OverrideDates]?: boolean }
```

Comment:
Instead of performing null checks later on (e.g., `preserveDate ?? false` or `skipMerging?.lockAt`), it will be easier to assign default values for them (`false` for `preserveDate` and `{}` for `skipMerging`).